Using 4.23.0, my app fails to start because of invalid checksums

### Search first

- [X] I searched and no similar issues were found

### Description

Our application (Java/Spring/MySQL) is currently using 4.20.0 and works fine. When I upgrade to 4.23.0, it fails to start due to Liquibase failing due to invalid checksums.

```
	Caused by: liquibase.exception.CommandExecutionException: liquibase.exception.ValidationFailedException: Validation Failed:
     1 changesets check sum
          org/foo/db/phase5/foo-bar-2017-09-14.xml::DECREASE_FP_FIELD_FOR_PDP::jdoe was: 8:83871094f06fbeba7a3755bb6e7980c0 but is now: 8:7afe9a1d89ce253fbca2d513dcdc2834
```

### Steps To Reproduce

The offending changeset (names changed to protect the innocent):

```xml
    <changeSet author="jdoe" id="DECREASE_FP_FIELD_FOR_PDP" context="bootstrap,demo,unit">
        <comment>Words words words. Please review if your institution would like this change. See JIRA-123 for details.</comment>

        <sql dbms="mysql">
            update FP_DV_PAYEE_DTL_T A join FP_DV_PAYEE_DTL_T B ON A.FDOC_NBR = B.FDOC_NBR set A.DV_RMT_PRSN_NM = SUBSTR(B.DV_RMT_PRSN_NM,1,75);
        </sql>

        <sql dbms="oracle">
            update FP_DV_PAYEE_DTL_T A set DV_RMT_PRSN_NM = (select SUBSTR(DV_RMT_PRSN_NM,1,75) from FP_DV_PAYEE_DTL_T B where A.FDOC_NBR = B.FDOC_NBR);
        </sql>

        <modifyDataType columnName="DV_RMT_PRSN_NM" newDataType="VARCHAR(75)" tableName="FP_DV_PAYEE_DTL_T"/>
    </changeSet>
```

### Expected/Desired Behavior

My app would startup like it does with 4.20.0.

### Liquibase Version

4.23.0

### Database Vendor & Version

MySQL 5.7

### Liquibase Integration

Spring (not Boot)

### Liquibase Extensions

_No response_

### OS and/or Infrastructure Type/Provider

macOS

### Additional Context

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR (Thank you!)

--- a/liquibase-integration-tests/src/test/groovy/liquibase/command/core/UpgradeChecksumVersionIntegrationTest.groovy
+++ b/liquibase-integration-tests/src/test/groovy/liquibase/command/core/UpgradeChecksumVersionIntegrationTest.groovy
@@ -19,16 +19,16 @@ import spock.lang.Unroll
 @LiquibaseIntegrationTest
 class UpgradeChecksumVersionIntegrationTest extends Specification{
     @Shared
-    private DatabaseTestSystem mysql = (DatabaseTestSystem) Scope.getCurrentScope().getSingleton(TestSystemFactory.class).getTestSystem("mysql")
+    private DatabaseTestSystem h2 = (DatabaseTestSystem) Scope.getCurrentScope().getSingleton(TestSystemFactory.class).getTestSystem("h2")
 
     private List<RanChangeSet> getRanChangesets(ChangeLogHistoryService changeLogService) {
-        mysql.getConnection().commit()
+        h2.getConnection().commit()
         changeLogService.reset()
         return changeLogService.getRanChangeSets()
     }
 
     private insertDbclRecord(String changelogfile) {
-        mysql.execute(new RawSqlStatement("""
+        h2.execute(new RawSqlStatement("""
 INSERT INTO DATABASECHANGELOG
 (ID, AUTHOR, FILENAME, DATEEXECUTED, ORDEREXECUTED, EXECTYPE, MD5SUM, DESCRIPTION, COMMENTS, TAG, LIQUIBASE, CONTEXTS, LABELS, DEPLOYMENT_ID)
 VALUES('1', 'your.name', '$changelogfile', '2023-05-31 14:33:39.108', 1, 'EXECUTED', '8:d925207397a5bb8863a41d513a65afd1', 'sql', 'example comment', NULL, 'DEV', 'example-context1', 'example-label', '5561619071');
@@ -38,7 +38,7 @@ VALUES('1', 'your.name', '$changelogfile', '2023-05-31 14:33:39.108', 1, 'EXECUT
     @Unroll
     def "update command should upgrade all checksums when no filters supplied" () {
         def changesetFilepath = "changelogs/common/checksum-changelog.xml"
-        final ChangeLogHistoryService changeLogService = ChangeLogHistoryServiceFactory.getInstance().getChangeLogService(mysql.getDatabaseFromFactory())
+        final ChangeLogHistoryService changeLogService = Scope.getCurrentScope().getSingleton(ChangeLogHistoryServiceFactory.class).getChangeLogService(h2.getDatabaseFromFactory())
         changeLogService.init()
         insertDbclRecord(storedFilepathPrefix + changesetFilepath)
 
@@ -51,9 +51,9 @@ VALUES('1', 'your.name', '$changelogfile', '2023-05-31 14:33:39.108', 1, 'EXECUT
 
         when:
         CommandScope updateCommandScope = new CommandScope(UpdateCommandStep.COMMAND_NAME)
-        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.URL_ARG, mysql.getConnectionUrl())
-        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.USERNAME_ARG, mysql.getUsername())
-        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.PASSWORD_ARG, mysql.getPassword())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.URL_ARG, h2.getConnectionUrl())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.USERNAME_ARG, h2.getUsername())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.PASSWORD_ARG, h2.getPassword())
         updateCommandScope.addArgumentValue(UpdateCommandStep.CHANGELOG_FILE_ARG, changesetFilepath)
         updateCommandScope.execute()
         ranChangeSets = getRanChangesets(changeLogService)
@@ -63,7 +63,7 @@ VALUES('1', 'your.name', '$changelogfile', '2023-05-31 14:33:39.108', 1, 'EXECUT
         ranChangeSets.forEach({ rcs -> assert rcs.getLastCheckSum().getVersion() == 9 })
 
         cleanup:
-        CommandUtil.runDropAll(mysql)
+        CommandUtil.runDropAll(h2)
 
         where:
         storedFilepathPrefix | _
@@ -73,7 +73,7 @@ VALUES('1', 'your.name', '$changelogfile', '2023-05-31 14:33:39.108', 1, 'EXECUT
 
     def "update command should upgrade only matching changesets when filter is applied" () {
         def changesetFilepath = "changelogs/common/checksum-changelog.xml"
-        final ChangeLogHistoryService changeLogService = ChangeLogHistoryServiceFactory.getInstance().getChangeLogService(mysql.getDatabaseFromFactory())
+        final ChangeLogHistoryService changeLogService = Scope.getCurrentScope().getSingleton(ChangeLogHistoryServiceFactory.class).getChangeLogService(h2.getDatabaseFromFactory())
         changeLogService.init()
         insertDbclRecord(changesetFilepath)
 
@@ -86,9 +86,9 @@ VALUES('1', 'your.name', '$changelogfile', '2023-05-31 14:33:39.108', 1, 'EXECUT
 
         when:
         CommandScope updateCommandScope = new CommandScope(UpdateCommandStep.COMMAND_NAME)
-        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.URL_ARG, mysql.getConnectionUrl())
-        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.USERNAME_ARG, mysql.getUsername())
-        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.PASSWORD_ARG, mysql.getPassword())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.URL_ARG, h2.getConnectionUrl())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.USERNAME_ARG, h2.getUsername())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.PASSWORD_ARG, h2.getPassword())
         updateCommandScope.addArgumentValue(UpdateCommandStep.CHANGELOG_FILE_ARG, changesetFilepath)
         updateCommandScope.addArgumentValue(UpdateCommandStep.CONTEXTS_ARG, "example-context2")
         updateCommandScope.execute()
@@ -100,12 +100,12 @@ VALUES('1', 'your.name', '$changelogfile', '2023-05-31 14:33:39.108', 1, 'EXECUT
         ranChangeSets.get(1).getLastCheckSum().getVersion() == 9
 
         cleanup:
-        CommandUtil.runDropAll(mysql)
+        CommandUtil.runDropAll(h2)
     }
 
     def "update-to-tag" () {
         def changesetFilepath = "changelogs/common/checksum-changelog-tag.xml"
-        final ChangeLogHistoryService changeLogService = ChangeLogHistoryServiceFactory.getInstance().getChangeLogService(mysql.getDatabaseFromFactory())
+        final ChangeLogHistoryService changeLogService = Scope.getCurrentScope().getSingleton(ChangeLogHistoryServiceFactory.class).getChangeLogService(h2.getDatabaseFromFactory())
         changeLogService.init()
         insertDbclRecord(changesetFilepath)
 
@@ -118,9 +118,9 @@ VALUES('1', 'your.name', '$changelogfile', '2023-05-31 14:33:39.108', 1, 'EXECUT
 
         when:
         CommandScope updateCommandScope = new CommandScope(UpdateToTagCommandStep.COMMAND_NAME)
-        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.URL_ARG, mysql.getConnectionUrl())
-        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.USERNAME_ARG, mysql.getUsername())
-        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.PASSWORD_ARG, mysql.getPassword())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.URL_ARG, h2.getConnectionUrl())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.USERNAME_ARG, h2.getUsername())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.PASSWORD_ARG, h2.getPassword())
         updateCommandScope.addArgumentValue(UpdateToTagCommandStep.CHANGELOG_FILE_ARG, changesetFilepath)
         updateCommandScope.addArgumentValue(UpdateToTagCommandStep.TAG_ARG, "version_1.3")
         updateCommandScope.execute()
@@ -133,12 +133,12 @@ VALUES('1', 'your.name', '$changelogfile', '2023-05-31 14:33:39.108', 1, 'EXECUT
         ranChangeSets.get(2).getLastCheckSum().getVersion() == 9
 
         cleanup:
-        CommandUtil.runDropAll(mysql)
+        CommandUtil.runDropAll(h2)
     }
 
     def "update-count"() {
         def changesetFilepath = "changelogs/common/checksum-changelog.xml"
-        final ChangeLogHistoryService changeLogService = ChangeLogHistoryServiceFactory.getInstance().getChangeLogService(mysql.getDatabaseFromFactory())
+        final ChangeLogHistoryService changeLogService = Scope.getCurrentScope().getSingleton(ChangeLogHistoryServiceFactory.class).getChangeLogService(h2.getDatabaseFromFactory())
         changeLogService.init()
         insertDbclRecord(changesetFilepath)
 
@@ -151,9 +151,9 @@ VALUES('1', 'your.name', '$changelogfile', '2023-05-31 14:33:39.108', 1, 'EXECUT
 
         when:
         CommandScope updateCommandScope = new CommandScope(UpdateCountCommandStep.COMMAND_NAME)
-        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.URL_ARG, mysql.getConnectionUrl())
-        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.USERNAME_ARG, mysql.getUsername())
-        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.PASSWORD_ARG, mysql.getPassword())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.URL_ARG, h2.getConnectionUrl())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.USERNAME_ARG, h2.getUsername())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.PASSWORD_ARG, h2.getPassword())
         updateCommandScope.addArgumentValue(UpdateCountCommandStep.CHANGELOG_FILE_ARG, changesetFilepath)
         updateCommandScope.addArgumentValue(UpdateCountCommandStep.COUNT_ARG, 1)
         updateCommandScope.execute()
@@ -165,16 +165,16 @@ VALUES('1', 'your.name', '$changelogfile', '2023-05-31 14:33:39.108', 1, 'EXECUT
         ranChangeSets.get(1).getLastCheckSum().getVersion() == 9
 
         cleanup:
-        CommandUtil.runDropAll(mysql)
+        CommandUtil.runDropAll(h2)
     }
 
 
     def "update-testing-rollback"() {
         def changesetFilepath = "changelogs/common/checksum-changelog.xml"
-        final ChangeLogHistoryService changeLogService = ChangeLogHistoryServiceFactory.getInstance().getChangeLogService(mysql.getDatabaseFromFactory())
+        final ChangeLogHistoryService changeLogService = Scope.getCurrentScope().getSingleton(ChangeLogHistoryServiceFactory.class).getChangeLogService(h2.getDatabaseFromFactory())
         changeLogService.init()
         insertDbclRecord(changesetFilepath)
-        mysql.executeSql("CREATE TABLE person(id int)")
+        h2.executeSql("CREATE TABLE person(id int)")
 
         when:
         def ranChangeSets = getRanChangesets(changeLogService)
@@ -185,9 +185,9 @@ VALUES('1', 'your.name', '$changelogfile', '2023-05-31 14:33:39.108', 1, 'EXECUT
 
         when:
         CommandScope updateTestingRollbackCommand = new CommandScope(UpdateTestingRollbackCommandStep.COMMAND_NAME)
-        updateTestingRollbackCommand.addArgumentValue(DbUrlConnectionCommandStep.URL_ARG, mysql.getConnectionUrl())
-        updateTestingRollbackCommand.addArgumentValue(DbUrlConnectionCommandStep.USERNAME_ARG, mysql.getUsername())
-        updateTestingRollbackCommand.addArgumentValue(DbUrlConnectionCommandStep.PASSWORD_ARG, mysql.getPassword())
+        updateTestingRollbackCommand.addArgumentValue(DbUrlConnectionCommandStep.URL_ARG, h2.getConnectionUrl())
+        updateTestingRollbackCommand.addArgumentValue(DbUrlConnectionCommandStep.USERNAME_ARG, h2.getUsername())
+        updateTestingRollbackCommand.addArgumentValue(DbUrlConnectionCommandStep.PASSWORD_ARG, h2.getPassword())
         updateTestingRollbackCommand.addArgumentValue(DatabaseChangelogCommandStep.CHANGELOG_FILE_ARG, changesetFilepath)
         updateTestingRollbackCommand.execute()
         ranChangeSets = getRanChangesets(changeLogService)
@@ -197,13 +197,13 @@ VALUES('1', 'your.name', '$changelogfile', '2023-05-31 14:33:39.108', 1, 'EXECUT
         ranChangeSets.forEach({ rcs -> assert rcs.getLastCheckSum().getVersion() == 9 })
 
         cleanup:
-        CommandUtil.runDropAll(mysql)
+        CommandUtil.runDropAll(h2)
     }
 
 
     def "update-sql calculates checksum but does not update DBCL"() {
         def changesetFilepath = "changelogs/common/checksum-changelog.xml"
-        final ChangeLogHistoryService changeLogService = ChangeLogHistoryServiceFactory.getInstance().getChangeLogService(mysql.getDatabaseFromFactory())
+        final ChangeLogHistoryService changeLogService = Scope.getCurrentScope().getSingleton(ChangeLogHistoryServiceFactory.class).getChangeLogService(h2.getDatabaseFromFactory())
         changeLogService.init()
         insertDbclRecord(changesetFilepath)
 
@@ -216,9 +216,9 @@ VALUES('1', 'your.name', '$changelogfile', '2023-05-31 14:33:39.108', 1, 'EXECUT
 
         when:
         CommandScope updateCommandScope = new CommandScope(UpdateSqlCommandStep.COMMAND_NAME)
-        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.URL_ARG, mysql.getConnectionUrl())
-        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.USERNAME_ARG, mysql.getUsername())
-        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.PASSWORD_ARG, mysql.getPassword())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.URL_ARG, h2.getConnectionUrl())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.USERNAME_ARG, h2.getUsername())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.PASSWORD_ARG, h2.getPassword())
         updateCommandScope.addArgumentValue(UpdateSqlCommandStep.CHANGELOG_FILE_ARG, changesetFilepath)
         updateCommandScope.execute()
         ranChangeSets = getRanChangesets(changeLogService)
@@ -228,12 +228,12 @@ VALUES('1', 'your.name', '$changelogfile', '2023-05-31 14:33:39.108', 1, 'EXECUT
         ranChangeSets.forEach({ rcs -> assert rcs.getLastCheckSum().getVersion() == 8 })
 
         cleanup:
-        CommandUtil.runDropAll(mysql)
+        CommandUtil.runDropAll(h2)
     }
 
     def "update-count-sql calculates checksum but does not update DBCL"() {
         def changesetFilepath = "changelogs/common/checksum-changelog.xml"
-        final ChangeLogHistoryService changeLogService = ChangeLogHistoryServiceFactory.getInstance().getChangeLogService(mysql.getDatabaseFromFactory())
+        final ChangeLogHistoryService changeLogService = Scope.getCurrentScope().getSingleton(ChangeLogHistoryServiceFactory.class).getChangeLogService(h2.getDatabaseFromFactory())
         changeLogService.init()
         insertDbclRecord(changesetFilepath)
 
@@ -246,9 +246,9 @@ VALUES('1', 'your.name', '$changelogfile', '2023-05-31 14:33:39.108', 1, 'EXECUT
 
         when:
         CommandScope updateCommandScope = new CommandScope(UpdateCountSqlCommandStep.COMMAND_NAME)
-        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.URL_ARG, mysql.getConnectionUrl())
-        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.USERNAME_ARG, mysql.getUsername())
-        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.PASSWORD_ARG, mysql.getPassword())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.URL_ARG, h2.getConnectionUrl())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.USERNAME_ARG, h2.getUsername())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.PASSWORD_ARG, h2.getPassword())
         updateCommandScope.addArgumentValue(UpdateCountSqlCommandStep.CHANGELOG_FILE_ARG, changesetFilepath)
         updateCommandScope.addArgumentValue(UpdateCountSqlCommandStep.COUNT_ARG, 1)
         updateCommandScope.execute()
@@ -259,12 +259,12 @@ VALUES('1', 'your.name', '$changelogfile', '2023-05-31 14:33:39.108', 1, 'EXECUT
         ranChangeSets.forEach({ rcs -> assert rcs.getLastCheckSum().getVersion() == 8 })
 
         cleanup:
-        CommandUtil.runDropAll(mysql)
+        CommandUtil.runDropAll(h2)
     }
 
     def "update-to-tag-sql calculates checksum but does not update DBCL"() {
         def changesetFilepath = "changelogs/common/checksum-changelog-tag.xml"
-        final ChangeLogHistoryService changeLogService = ChangeLogHistoryServiceFactory.getInstance().getChangeLogService(mysql.getDatabaseFromFactory())
+        final ChangeLogHistoryService changeLogService = Scope.getCurrentScope().getSingleton(ChangeLogHistoryServiceFactory.class).getChangeLogService(h2.getDatabaseFromFactory())
         changeLogService.init()
         insertDbclRecord(changesetFilepath)
 
@@ -277,9 +277,9 @@ VALUES('1', 'your.name', '$changelogfile', '2023-05-31 14:33:39.108', 1, 'EXECUT
 
         when:
         CommandScope updateCommandScope = new CommandScope(UpdateToTagSqlCommandStep.COMMAND_NAME)
-        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.URL_ARG, mysql.getConnectionUrl())
-        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.USERNAME_ARG, mysql.getUsername())
-        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.PASSWORD_ARG, mysql.getPassword())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.URL_ARG, h2.getConnectionUrl())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.USERNAME_ARG, h2.getUsername())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.PASSWORD_ARG, h2.getPassword())
         updateCommandScope.addArgumentValue(UpdateToTagSqlCommandStep.CHANGELOG_FILE_ARG, changesetFilepath)
         updateCommandScope.addArgumentValue(UpdateToTagSqlCommandStep.TAG_ARG, "version_1.3")
         updateCommandScope.execute()
@@ -290,6 +290,49 @@ VALUES('1', 'your.name', '$changelogfile', '2023-05-31 14:33:39.108', 1, 'EXECUT
         ranChangeSets.forEach({ rcs -> assert rcs.getLastCheckSum().getVersion() == 8 })
 
         cleanup:
-        CommandUtil.runDropAll(mysql)
+        CommandUtil.runDropAll(h2)
+    }
+
+    @Unroll
+    def "DBMS filtered changes should be used to calculate v8 checksum but not v9" () {
+        def changesetFilepath = "changelogs/h2/checksum/dbms-filter-changelog.xml"
+        final ChangeLogHistoryService changeLogService = Scope.getCurrentScope().getSingleton(ChangeLogHistoryServiceFactory.class).getChangeLogService(h2.getDatabaseFromFactory())
+        changeLogService.init()
+        h2.execute(new RawSqlStatement("""
+INSERT INTO DATABASECHANGELOG
+(ID, AUTHOR, FILENAME, DATEEXECUTED, ORDEREXECUTED, EXECTYPE, MD5SUM, DESCRIPTION, COMMENTS, TAG, LIQUIBASE, CONTEXTS, LABELS, DEPLOYMENT_ID)
+VALUES('1', 'fl', '$changesetFilepath', '2023-09-29 14:33:39.108', 1, 'EXECUTED', '8:0a36c7b201a287dd3348e8dd19e44be7', 'sql', 'example comment', NULL, 'DEV', 'example-context1', 'example-label', '5561619071');
+"""))
+        h2.execute(new RawSqlStatement("""
+INSERT INTO DATABASECHANGELOG
+(ID, AUTHOR, FILENAME, DATEEXECUTED, ORDEREXECUTED, EXECTYPE, MD5SUM, DESCRIPTION, COMMENTS, TAG, LIQUIBASE, CONTEXTS, LABELS, DEPLOYMENT_ID)
+VALUES('2', 'fl', '$changesetFilepath', '2023-09-29 14:33:39.112', 2, 'EXECUTED', '8:a6a54dbc65048ebf1388da78c31ef1a9', 'sqlFile; sqlFile', '', NULL, 'DEV', 'example-context1', 'example-label', '5561619071');
+"""))
+
+        when:
+        def ranChangeSets = getRanChangesets(changeLogService)
+
+        then:
+        ranChangeSets.size() == 2
+        ranChangeSets.forEach({ rcs -> assert rcs.getLastCheckSum().getVersion() == 8 })
+
+        when:
+        CommandScope updateCommandScope = new CommandScope(UpdateCommandStep.COMMAND_NAME)
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.URL_ARG, h2.getConnectionUrl())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.USERNAME_ARG, h2.getUsername())
+        updateCommandScope.addArgumentValue(DbUrlConnectionCommandStep.PASSWORD_ARG, h2.getPassword())
+        updateCommandScope.addArgumentValue(UpdateCommandStep.CHANGELOG_FILE_ARG, changesetFilepath)
+        updateCommandScope.execute()
+        ranChangeSets = getRanChangesets(changeLogService)
+
+        then:
+        ranChangeSets.size() == 3
+        ranChangeSets.get(0).getLastCheckSum().toString() == "9:d41d8cd98f00b204e9800998ecf8427e"
+        ranChangeSets.get(1).getLastCheckSum().toString() == "9:62336a615e62e89c9d86128d1ca60ecf"
+        ranChangeSets.get(2).getLastCheckSum().toString() == "9:3b78b9910981c62ca27148640fad7bc2"
+
+        cleanup:
+        CommandUtil.runDropAll(h2)
+
     }
 }

--- a/liquibase-integration-tests/src/test/resources/changelogs/h2/checksum/create-table.sql
+++ b/liquibase-integration-tests/src/test/resources/changelogs/h2/checksum/create-table.sql
@@ -1,0 +1,1 @@
+create table thisShouldRun(id int)

--- a/liquibase-integration-tests/src/test/resources/changelogs/h2/checksum/dbms-filter-changelog.xml
+++ b/liquibase-integration-tests/src/test/resources/changelogs/h2/checksum/dbms-filter-changelog.xml
@@ -1,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<databaseChangeLog
+        xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xmlns:pro="http://www.liquibase.org/xml/ns/pro"
+        xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-latest.xsd
+    http://www.liquibase.org/xml/ns/pro http://www.liquibase.org/xml/ns/pro/liquibase-pro-latest.xsd ">
+
+    <changeSet author="fl" id="1">
+        <sql dbms="thisDbWillNeverExist">
+                I can write anything here as this content will only be used
+                  to generate checksum on versions before 9 </sql>
+    </changeSet>
+
+    <changeSet author="fl" id="2">
+        <sqlFile dbms="thisDbWillNeverExist" relativeToChangelogFile="true" path="create-table.sql" />
+        <sqlFile dbms="h2" relativeToChangelogFile="true" path="create-table.sql" />
+    </changeSet>
+
+    <changeSet author="fl" id="3">
+            <!-- this will run succesfully too -->
+        <sql dbms="h2">select 1</sql>
+    </changeSet>
+
+</databaseChangeLog>

--- a/liquibase-standard/src/main/java/liquibase/changelog/ChangeLogIterator.java
+++ b/liquibase-standard/src/main/java/liquibase/changelog/ChangeLogIterator.java
@@ -4,7 +4,6 @@ import liquibase.ContextExpression;
 import liquibase.Labels;
 import liquibase.RuntimeEnvironment;
 import liquibase.Scope;
-import liquibase.change.CheckSum;
 import liquibase.changelog.filter.ChangeSetFilter;
 import liquibase.changelog.filter.ChangeSetFilterResult;
 import liquibase.changelog.visitor.ChangeSetVisitor;
@@ -97,7 +96,12 @@ public class ChangeLogIterator {
                         }
 
                         boolean finalShouldVisit = shouldVisit;
-                        Scope.child(Scope.Attr.changeSet.name(), changeSet, () -> {
+
+                        Map<String, Object> scopeValues = new HashMap<>();
+                        scopeValues.put(Scope.Attr.changeSet.name(), changeSet);
+                        scopeValues.put(Scope.Attr.database.name(), env.getTargetDatabase());
+
+                        Scope.child(scopeValues, () -> {
                             if (finalShouldVisit) {
                                 //
                                 // Go validate any changesets with an Executor if

--- a/liquibase-standard/src/main/java/liquibase/changelog/ChangeSet.java
+++ b/liquibase-standard/src/main/java/liquibase/changelog/ChangeSet.java
@@ -737,20 +737,22 @@ public class ChangeSet implements Conditional, ChangeLogChild {
 
                 log.fine("Reading ChangeSet: " + this);
                 for (Change change : getChanges()) {
-                    if (listener != null) {
-                        listener.willRun(change, this, changeLog, database);
-                    }
-                    if (change.generateStatementsVolatile(database)) {
-                        executor.comment("WARNING The following SQL may change each run and therefore is possibly incorrect and/or invalid:");
-                    }
+                    if ((!(change instanceof DbmsTargetedChange)) || DatabaseList.definitionMatches(((DbmsTargetedChange) change).getDbms(), database, true)) {
+                        if (listener != null) {
+                            listener.willRun(change, this, changeLog, database);
+                        }
+                        if (change.generateStatementsVolatile(database)) {
+                            executor.comment("WARNING The following SQL may change each run and therefore is possibly incorrect and/or invalid:");
+                        }
 
-                    String sql = addSqlMdc(change, database, false);
-                    this.getGeneratedSql().add(sql);
+                        String sql = addSqlMdc(change, database, false);
+                        this.getGeneratedSql().add(sql);
 
-                    database.executeStatements(change, databaseChangeLog, sqlVisitors);
-                    log.info(change.getConfirmationMessage());
-                    if (listener != null) {
-                        listener.ran(change, this, changeLog, database);
+                        database.executeStatements(change, databaseChangeLog, sqlVisitors);
+                        log.info(change.getConfirmationMessage());
+                        if (listener != null) {
+                            listener.ran(change, this, changeLog, database);
+                        }
                     } else {
                         log.fine("Change " + change.getSerializedObjectName() + " not included for database " + database.getShortName());
                     }
@@ -1154,7 +1156,7 @@ public class ChangeSet implements Conditional, ChangeLogChild {
     }
 
     public String toString(boolean includeMD5Sum) {
-        ChecksumVersion checksumVersion = ChecksumVersion.enumFromChecksumVersion(this.checkSum != null ? this.checkSum.getVersion() : CheckSum.getCurrentVersion());
+        ChecksumVersion checksumVersion = ChecksumVersion.enumFromChecksumVersion(this.checkSum != null ? this.checkSum.getVersion() : ChecksumVersion.latest().getVersion());
         return filePath + "::" + getId() + "::" + getAuthor() +
                 (includeMD5Sum ? ("::(Checksum: " + generateCheckSum(checksumVersion) + ")") : "");
     }

--- a/liquibase-standard/src/main/java/liquibase/changelog/ChangeSet.java
+++ b/liquibase-standard/src/main/java/liquibase/changelog/ChangeSet.java
@@ -363,7 +363,11 @@ public class ChangeSet implements Conditional, ChangeLogChild {
                 if (checkSum == null) {
                     StringBuilder stringToMD5 = new StringBuilder();
                     for (Change change : this.getChanges()) {
-                        stringToMD5.append(change.generateCheckSum()).append(":");
+                        if (Scope.getCurrentScope().getChecksumVersion().lowerOrEqualThan(ChecksumVersion.V8) ||
+                            !(change instanceof DbmsTargetedChange) ||
+                            DatabaseList.definitionMatches(((DbmsTargetedChange) change).getDbms(), Scope.getCurrentScope().getDatabase(), true)) {
+                            stringToMD5.append(change.generateCheckSum()).append(":");
+                        }
                     }
 
                     for (SqlVisitor visitor : this.getSqlVisitors()) {

--- a/liquibase-standard/src/main/java/liquibase/changelog/ChangeSet.java
+++ b/liquibase-standard/src/main/java/liquibase/changelog/ChangeSet.java
@@ -363,9 +363,11 @@ public class ChangeSet implements Conditional, ChangeLogChild {
                 if (checkSum == null) {
                     StringBuilder stringToMD5 = new StringBuilder();
                     for (Change change : this.getChanges()) {
-                        if (Scope.getCurrentScope().getChecksumVersion().lowerOrEqualThan(ChecksumVersion.V8) ||
-                            !(change instanceof DbmsTargetedChange) ||
-                            DatabaseList.definitionMatches(((DbmsTargetedChange) change).getDbms(), Scope.getCurrentScope().getDatabase(), true)) {
+                        // checksum v8 requires changes that are applied even to other databases to be calculated
+                        // checksum v9 excludes them from calculation
+                        if (!(change instanceof DbmsTargetedChange) ||
+                                Scope.getCurrentScope().getChecksumVersion().lowerOrEqualThan(ChecksumVersion.V8) ||
+                                DatabaseList.definitionMatches(((DbmsTargetedChange) change).getDbms(), Scope.getCurrentScope().getDatabase(), true)) {
                             stringToMD5.append(change.generateCheckSum()).append(":");
                         }
                     }

--- a/liquibase-standard/src/main/java/liquibase/changelog/filter/DbmsChangeSetFilter.java
+++ b/liquibase-standard/src/main/java/liquibase/changelog/filter/DbmsChangeSetFilter.java
@@ -27,20 +27,12 @@ public class DbmsChangeSetFilter implements ChangeSetFilter {
             return new ChangeSetFilterResult(true, "No database connection, cannot evaluate dbms attribute", this.getClass(), getMdcName(), getDisplayName());
         }
         List<SqlVisitor> visitorsToRemove = new ArrayList<>();
-        List<Change> changesToRemove = new ArrayList<>();
         for (SqlVisitor visitor : changeSet.getSqlVisitors()) {
             if (!DatabaseList.definitionMatches(visitor.getApplicableDbms(), database, true)) {
                 visitorsToRemove.add(visitor);
             }
         }
-        for(Change change : changeSet.getChanges()){
-            if (((change instanceof DbmsTargetedChange)) && !DatabaseList.definitionMatches(((DbmsTargetedChange) change).getDbms(), database, true)){
-                changesToRemove.add(change);
-            }
-        }
-
         changeSet.getSqlVisitors().removeAll(visitorsToRemove);
-        changeSet.removeAllChanges(changesToRemove);
 
         String dbmsList;
         if ((changeSet.getDbmsSet() == null) || changeSet.getDbmsSet().isEmpty()) {

